src/protocol/sort/secureapplyinv.rs

```diff
@@ -0,0 +1,171 @@
+        for i in &permutation {
+            perms.push(i.as_u128() as usize);
+        }
+        let mut permutation = Permutation::oneline(perms);
+
+        // 5. apply the permutation on the shuffled input
+        // The paper suggests taking apply_inv. However, Permutation::oneline returns an inverted permutation, so we are negating
+        // the effect by just calling apply instead of apply_inv here
+        apply(&mut permutation, &mut self.input);
+        Ok(())
+    }
+
+    #[allow(clippy::cast_possible_truncation)]
+    async fn reveal_permutations<N: Network>(
+        &self,
+        ctx: &ProtocolContext<'_, N>,
+    ) -> Result<Vec<F>, BoxError> {
+        let reveals = self
+            .permutation
+            .iter()
+            .enumerate()
+            .map(|(index, input)| async move {
+                reveal(
+                    ctx.narrow(&index.to_string()),
+                    RecordId::from(index as u32),
+                    *input,
+                )
+                .await
+            });
+        try_join_all(reveals).await
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use permutation::Permutation;
+    use rand::seq::SliceRandom;
+    use tokio::try_join;
+
+    use crate::{
+        field::Fp31,
+        protocol::{sort::apply::apply_inv, QueryId},
+        test_fixture::{generate_shares, make_contexts, make_world, validate_and_reconstruct},
+    };
+
+    use super::SecureApplyInv;
+
+    #[tokio::test]
+    pub async fn secureapplyinv() {
+        const BATCHSIZE: usize = 25;
+        for _ in 0..10 {
+            let mut rng = rand::thread_rng();
+            let mut input = [0u128; BATCHSIZE];
+            random_number::random_fill!(input, 0..24);
+
+            let mut permutation: Vec<usize> = (0..BATCHSIZE).collect();
+            permutation.shuffle(&mut rng);
+
+            let mut expected_result = input;
+            let mut cloned_perm = Permutation::oneline(permutation.clone()).inverse();
+
+            apply_inv(&mut cloned_perm, &mut expected_result);
+
+            let permutation: Vec<u128> = permutation.iter().map(|x| *x as u128).collect();
+
+            let mut perm_shares = generate_shares(permutation);
+            let mut input_shares = generate_shares(input.to_vec());
+
+            let world = make_world(QueryId);
+            let context = make_contexts(&world);
+
+            let mut secure_applyinv0 = SecureApplyInv::new(&mut input_shares.0, &mut perm_shares.0);
+            let mut secure_applyinv1 = SecureApplyInv::new(&mut input_shares.1, &mut perm_shares.1);
+            let mut secure_applyinv2 = SecureApplyInv::new(&mut input_shares.2, &mut perm_shares.2);
+
+            let h0_future = secure_applyinv0.execute(&context[0]);
+            let h1_future = secure_applyinv1.execute(&context[1]);
+            let h2_future = secure_applyinv2.execute(&context[2]);
+
+            try_join!(h0_future, h1_future, h2_future).unwrap();
```

Comment:
I'm using this syntax (that I like better). It avoids macros and so many local variables:

```
let results = try_join_all(vec![
    SecureApplyInv::new(...)
        .execute(...),
    SecureApplyInv::new(...)
        .execute(...),
    SecureApplyInv::new(...)
        .execute(...),
]).await?;

Reply:
Love it! I am going to change all the protocols to have this syntax. For now, only changing SecureApplyInv.

Reply:
Before we batch change all of them - let's discuss. I too suggested this and @akoshelev had the idea that we would use the ::new() method to pass all of the *unique* arguments, while keeping `execute()` constant, i.e. always accepting the ProtocolContext and RecordId.

Reply:
I started making the changes in Shuffle (acted too fast). Reverting them for now

Reply:
:). Let’s land this PR first. It’s really close.

Reply:
Yep, I added this topic to our agenda for Nov offsite